api/session.ts

```diff
@@ -0,0 +1,429 @@
+						if (ladders) {
+							userType = '4';
+							void users.update(userid, {banstate: -10});
+						}
+					}
+				}
+			}
+			const logintime = userData.logintime;
+			if (!logintime || time() - logintime > LOGINTIME_INTERVAL) {
+				await users.update(userid, {logintime: time(), loginip: ip});
+			}
+			data = user.id + ',' + userType + ',' + time() + ',' + serverHost;
+		} else {
+			if (userid.length < 1 || !/[a-z]/.test(userid)) {
+				return ';;Your username must contain at least one letter.';
+			}
+			const userstate = await users.get('*', userid);
+			if (userstate) {
+				if (userstate.banstate >= 100 || ((userstate as any).password && userstate.nonce)) {
+					return ';;Your username is no longer available.';
+				}
+				if (userstate.email?.endsWith('@')) {
+					return ';;@gmail';
+				}
+				return ';';
+			} else {
+				// Unregistered username.
+				userType = '1';
+				if (forceUsertype) userType = forceUsertype;
+				data = user + ',' + userType + ',' + time() + ',' + serverHost;
+				this.updateCookie();
+			}
+		}
+		let splitChallenge: string[] = [];
+		for (const delim of [';', '%7C', '|']) {
+			splitChallenge = challenge.split(delim);
+			if (splitChallenge.length > 1) break;
+		}
+		let challengetoken;
+		if (splitChallenge.length > 1) {
+			challengekeyid = parseInt(splitChallenge[0], 10);
+			challenge = splitChallenge[1];
+			if (splitChallenge[2] && !challengetoken) challengetoken = splitChallenge[2];
+		}
+
+		if (!toID(challenge)) {
+			// Bogus challenge.
+			return ';;Corrupt challenge';
+		}
+		if (challengekeyid < 1) {
+			return ';;This server is requesting an invalid login key. This probably means that either you are not connected to a server, or the server is set up incorrectly.';
+		} else if (Config.compromisedkeys?.includes(challengekeyid)) {
+			// Compromised keys - no longer supported.
+			return `;;This server is using login key ${challengekeyid}, which is no longer supported. Please tell the server operator to update their config.js file.`;
+		} else if (!Config.privatekeys?.[challengekeyid]) {
+			// Bogus key id.
+			return ';;Unknown key ID';
+		} else {
+			// Include the challenge in the assertion.
+			data = (challengeprefix || "") + challenge + ',' + data;
+		}
+		this.updateCookie();
+
+		if (Config.validateassertion) {
+			data = Config.validateassertion.call(this, data, user, ip, server);
+		}
+
+		const sign = crypto.createSign('RSA-SHA1');
+		sign.update(data);
+		sign.end();
+		const sig = sign.sign(Config.privatekeys[challengekeyid], 'hex');
+		return data + ';' + sig;
+	}
+	static isUseridAllowed(userid: string) {
+		const disallowed = [
+			...(Config.bannedTerms || []),
+			'nigger', 'nigga', 'faggot',
+			'lolicon', 'roricon', 'lazyafrican',
+			'tranny',
+		];
+		for (const term of disallowed) {
+			if (userid.includes(term)) return false;
+		}
+		return true;
+	}
+	static wordfilter(user: string) {
```

Comment:
What's the difference between this and `isUseridAllowed`?

Reply:
It replaces the banned terms with a *. Wordfilter is used to censor ladders and replays.

Reply:
OK. It seems like we would want anything banned in a username to also be censored in replays, but that's probably for another PR. Thanks for explaining!